### PR TITLE
BUGFIX: improve performance of `Scripts::buildPhpCommand`

### PR DESCRIPTION
**Pull request by mhsdesign**
https://github.com/neos/flow-development-collection/pull/2491 seems to cause a tiny minor performance regression. Booting minimal flow (only configuration manager), and running Scripts::buildPhpCommand went from `0.025s` to `0.044s` on my super fast machine ... it might be a more measurable difference on slower systems.

using `exec` and invoking another php process is just slower than not to.

Following changes were made:

A: runtime cache the buildPhpCommand
B: remove the overhead of a second `exec` by using first native realpath and only fallback in case of an open base dir restriction to `exec`

see https://github.com/neos/flow-development-collection/pull/2491#issuecomment-1639964461

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Flow/Classes/Core/Booting/Scripts.php [hunk 1]
```diff
@@ -60,6 +60,8 @@
  */
 class Scripts
 {
+    protected static ?string $buildPhpCommand = null;
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Core/Booting/Scripts.php]:
```suggestion
    protected static ?string $builtPhpCommand = null;
```
as it _has been built_ if it's in there. 😎

File: Neos.Flow/Classes/Core/Booting/Scripts.php [hunk 2]
```diff
@@ -775,6 +777,10 @@ protected static function buildSubprocessCommand(string $commandIdentifier, arra
      */
     public static function buildPhpCommand(array $settings): string
     {
+        if (isset(static::$buildPhpCommand)) {
+            return static::$buildPhpCommand;
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Core/Booting/Scripts.php]:
```suggestion
        if (isset(static::$builtPhpCommand)) {
            return static::$builtPhpCommand;
```

File: Neos.Flow/Classes/Core/Booting/Scripts.php [hunk 3]
```diff
@@ -812,7 +818,7 @@ public static function buildPhpCommand(array $settings): string
 
         static::ensureWebSubrequestsUseCurrentlyRunningPhpVersion($command);
 
-        return $command;
+        return static::$buildPhpCommand = $command;
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Core/Booting/Scripts.php]:
```suggestion
        return static::$builtPhpCommand = $command;
```

**Review by kdambekalns — CHANGES_REQUESTED** [review]
Mind the tense! `buildPhpCommand` sounds like a flag or command, `builtPhpCommand` at least tells me something has been done already…

Also, the tests fail – might be a "test-only" side effect, but please investigate.

File: Neos.Flow/Classes/Core/Booting/Scripts.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Flow/Classes/Core/Booting/Scripts.php]: yes mister english teacher ;) ^^

**Comment by mhsdesign** [comment]
> Also, the tests fail – might be a "test-only" side effect, but please investigate.

all adjusted ;)

File: Neos.Flow/Classes/Core/Booting/Scripts.php [hunk 4]
```diff
@@ -60,6 +60,8 @@
  */
 class Scripts
 {
+    protected static ?string $builtPhpCommand = null;
```
Comment (mhsdesign) [inline Neos.Flow/Classes/Core/Booting/Scripts.php]:
```suggestion
    /** @var string */
    protected static $builtPhpCommand = null;
```

**Review by fcool — APPROVED** [review]

File: Neos.Flow/Classes/Core/Booting/Scripts.php [hunk 5]
```diff
@@ -849,8 +856,13 @@ protected static function ensureCLISubrequestsUseCurrentlyRunningPhpBinary($phpB
             return;
         }
 
-        exec(PHP_BINARY . ' -r "echo realpath(PHP_BINARY);"', $output);
-        $realPhpBinary = $output[0];
+        // stfu to avoid possible open_basedir restriction https://github.com/neos/flow-development-collection/pull/2491
+        $realPhpBinary = @realpath(PHP_BINARY);
```
Comment (mhsdesign) [inline Neos.Flow/Classes/Core/Booting/Scripts.php:860]: alternatively we can just check if `ini_get('open_basedir')` is not empty and then assume `realpath` will not work ...

**Review by kdambekalns — APPROVED** [review]

**Pull request merged**